prototypes/matrix_free_navier_stokes/matrix_free_navier_stokes.cc

```diff
@@ -0,0 +1,2420 @@
+
+  using VectorType = LinearAlgebra::distributed::Vector<number>;
+
+  using value_type = number;
+
+  using size_type = VectorizedArray<number>;
+
+  NavierStokesOperator();
+
+  NavierStokesOperator(const MappingQ<dim> &            mapping,
+                       const DoFHandler<dim> &          dof_handler,
+                       const AffineConstraints<number> &constraints,
+                       const QGauss<1> &                quadrature,
+                       const Settings &                 parameters,
+                       const unsigned int               mg_level);
+
+  void
+  reinit(const MappingQ<dim> &            mapping,
+         const DoFHandler<dim> &          dof_handler,
+         const AffineConstraints<number> &constraints,
+         const QGauss<1> &                quadrature,
+         const Settings &                 parameters,
+         const unsigned int               mg_level);
+
+  void
+  compute_element_size();
+
+  types::global_dof_index
+  m() const;
+
+  number
+  el(unsigned int, unsigned int) const;
+
+  void
+  clear();
+
+  void
+  initialize_dof_vector(VectorType &vec) const;
+
+  const std::shared_ptr<const Utilities::MPI::Partitioner> &
+  get_vector_partitioner() const;
+
+  void
+  vmult(VectorType &dst, const VectorType &src) const;
+
+  void
+  Tvmult(VectorType &dst, const VectorType &src) const;
+
+  void
+  vmult_interface_down(VectorType &dst, VectorType const &src) const;
+
+  void
+  vmult_interface_up(VectorType &dst, VectorType const &src) const;
+
+  const TrilinosWrappers::SparseMatrix &
+  get_system_matrix() const;
+
+  const MatrixFree<dim, number> &
+  get_system_matrix_free() const;
+
+  const AlignedVector<VectorizedArray<number>>
+  get_element_size() const;
+
+  void
+  compute_inverse_diagonal(VectorType &diagonal) const;
+
+  void
+  evaluate_non_linear_term(const VectorType &newton_step);
+
+private:
+  void
+  do_cell_integral_local(FECellIntegrator &integrator) const;
+
+  void
+  do_cell_integral_range(
+    const MatrixFree<dim, number> &              matrix_free,
+    VectorType &                                 dst,
+    const VectorType &                           src,
+    const std::pair<unsigned int, unsigned int> &range) const;
+
+  static IndexSet
+  get_refinement_edges(const MatrixFree<dim, number> &matrix_free);
+
+private:
+  MatrixFree<dim, number>                matrix_free;
+  AffineConstraints<number>              constraints;
+  mutable TrilinosWrappers::SparseMatrix system_matrix;
+  Settings                               parameters;
+  AlignedVector<VectorizedArray<number>> element_size;
+  Table<2, Tensor<1, dim + 1, VectorizedArray<number>>>
+    nonlinear_previous_values;
+  Table<2, Tensor<1, dim + 1, Tensor<1, dim, VectorizedArray<number>>>>
+    nonlinear_previous_gradient;
+  Table<2, Tensor<1, dim + 1, Tensor<1, dim, VectorizedArray<number>>>>
+    nonlinear_previous_hessian_diagonal;
```

Comment:
I would probably only store the global vector and recompute these quantities with `FEEvaluaion` (the Hessian is quite large).

Reply:
That's a good point. I will leave the prototype like this, but I will take this into account for the actual implementation in Lethe :)